packages/cinterop/src/commonMain/kotlin/io/realm/kotlin/internal/interop/SynchronizableObject.kt

```diff
@@ -0,0 +1,30 @@
+/*
+ * Copyright 2023 Realm Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.realm.kotlin.internal.interop
+
+/**
+ * A __synchronizable object__ that can be used to enforce mutual exclusion so that only one
+ * thread can execute the accompanied block across all [synchronized]-call for the same
+ * [SynchronizableObject] at a time.
+ */
+expect class SynchronizableObject()
+
+/**
+ * Execute the given `block` ensuring that no other [synchronized]-call is executing its block for
+ * the same [SynchronizableObject] at the same time.
+ */
+expect inline fun <R> synchronized(lock: SynchronizableObject, block: () -> R): R
```

Comment:
I saw the resemblance with the synchronized block 😃 

I find it easier to read if we invoke the function as an object method. It is not a big deal, but I like how extension functions allow the separation of the parameters of a function from the object it's invoked on.

Reply:
I don't have strong opinions on this, but writing `synchronized(obj)` also looked a bit funny to me (even though that is the Java syntax)...probably because it can only take a `SynchronizedObject`.

My immediate reaction was going for something like `Mutex().withLock {  }`.

Using `run` doesn't really convey the intent of what is happening IMO.

#namingIsHard

Reply:
Went for `SynchronizedObject.withLock { }`